src/scheduler/mod.rs

```diff
@@ -800,119 +814,114 @@ mod test {
-            Scheduler::<Rsrc>::new_with_poll_interval(
-                tempfile::tempdir().unwrap().into_path(),
-                Duration::from_millis(1),
-            )
-            .expect("Failed to create scheduler"),
-        );
+    fn is_sorted(v: &Vec<usize>) -> bool {
+        let mut sorted = v.clone();
+        sorted.sort();
+        *v == sorted
     }
 
+    new_scheduler!(SCHEDULER3);
     #[test]
-
     fn test_schedule_fn() {
-        let scheduler_handle = Scheduler::start(&*SCHEDULER3).expect("Failed to start scheduler.");
-        thread::sleep(Duration::from_millis(300));
+        use rand::seq::SliceRandom;
+
+        let scheduler = &*SCHEDULER3;
+        let handle = Scheduler::start(scheduler).unwrap();
+
+        let mut results = Arc::new(Mutex::new(Vec::new()));
+
+        const NUM_RESOURCES: usize = 1;
+        const NUM_TASKS: usize = 10;
+        let resources = (0..NUM_RESOURCES).map(|id| Rsrc { id }).collect::<Vec<_>>();
 
-        let n = 10;
-        let resources = (0..n).map(|id| Rsrc { id }).collect::<Vec<_>>();
-        let mut futures = (0..n)
+        let mut order = (0..NUM_TASKS).collect::<Vec<_>>();
+        order.shuffle(&mut thread_rng());
+
+        let mut futures = order
+            .into_iter()
             .map(|i| {
-                Some((*SCHEDULER3).lock().unwrap().schedule_future(
-                    n - i,
+                let results = Arc::clone(&results);
+                scheduler.lock().unwrap().schedule_future(
+                    i,
                     &format!("task {}", i),
                     &resources,
-                    move |_rsrc| i,
+                    move |rsrc| {
+                        thread::sleep(Duration::from_millis(200));
+                        results.lock().unwrap().push(i);
+                    },
                     false,
-                ))
+                )
             })
             .collect::<Vec<_>>();
 
-        // This only tests that every function is run and returns its expected value.
-        // The order of esults is determined by the order in which the task functions are scheduled,
-        // not on that in which they are performed. Actual scheduling is not exercised here.
-        let mut results = Vec::new();
-
-        thread::sleep(Duration::from_millis(100));
+        futures
+            .into_iter()
+            .for_each(|f| tokio_test::block_on(f).unwrap());
 
-        for future in futures.iter_mut() {
-            if let Some(future) = future.take() {
-                if let Some(value) = future.now_or_never() {
-                    results.push(value.unwrap());
-                }
-            }
-        }
+        let results = results.lock().unwrap().clone();
 
-        let mut expected = (0..n).collect::<Vec<_>>();
+        // First NUM_RESOURCES tasks might be misschedulized
+        let results = results[NUM_RESOURCES..].to_vec();
 
-        assert_eq!(expected, results);
+        assert!(is_sorted(&results));
```

Comment:
This test now seems to be doing something completely different. Whether or not the new one does something useful, I think this refactor should preserve the behavior of the previous.